pennylane/operation.py

```diff
@@ -300,48 +394,100 @@ def hash(self):
+            of a particular operator. By default, it should always take the
+            operator's parameters and hyperparameters as inputs.
 
-        **Example:**
+            Alternatively, a custom signature can be defined, in which case the `matrix()`
+            method has to be overwritten as well.
 
-        >>> qml.RY._matrix(0.5)
-        >>> array([[ 0.96891242+0.j, -0.24740396+0.j],
-                   [ 0.24740396+0.j,  0.96891242+0.j]])
+        Args:
+            params (list): trainable parameters of this operator, as stored in ``op.parameters``
+            hyperparams (dict): non-trainable hyperparameters of this operator, as stored in ``op.hyperparameters``
 
         Returns:
-            array: matrix representation
+            tensor_like or None: matrix representation
+
+        **Example**
+
+        >>> qml.CNOT.compute_matrix()
+        [[1, 0, 0, 0],
+        [0, 1, 0, 0],
+        [0, 0, 0, 1],
+        [0, 0, 1, 0]]
+
+        The matrix representation may depend on parameters or hyperparameters:
+
+        >>> qml.Rot.compute_matrix(0.1, 0.2, 0.3)
+        [[ 0.97517033-0.19767681j -0.09933467+0.00996671j]
+         [ 0.09933467+0.00996671j  0.97517033+0.19767681j]]
+
+        If parameters are tensors, a tensor of the same type is returned:
+
+        >>> res = qml.Rot.compute_matrix(torch.tensor(0.1), torch.tensor(0.2), torch.tensor(0.3))
+        >>> type(res)
+        <class 'torch.Tensor'>
         """
-        raise NotImplementedError
+        return None
 
-    @property
-    def matrix(self):
-        r"""Matrix representation of an instantiated operator
-        in the computational basis.
+    def matrix(self, wire_order=None):
+        r"""Matrix representation of this operator in the computational basis.
 
-        **Example:**
+        If ``wire_order`` is provided, the
+        numerical representation considers the position of the
+        operator's wires in the global wire order.
+        Otherwise, the wire order defaults to ``self.wires``.
 
-        >>> U = qml.RY(0.5, wires=1)
-        >>> U.matrix
-        >>> array([[ 0.96891242+0.j, -0.24740396+0.j],
-                   [ 0.24740396+0.j,  0.96891242+0.j]])
+        If the matrix depends on trainable parameters, the result
+        will be cast in the same autodifferentiation framework as the parameters.
+
+        .. note::
+            By default, this method calls the static method ``compute_matrix``,
+            which is used by subclasses to define the actual matrix representation.
+            The call assumes that the static method has the signature ``(*params, **hyperparams)``,
+            where ``params`` refers to ``op.parameters`` and ``hyperparams``
+            to ``op.hyperparameters``, and ``op`` is an instance of this operator.
+
+            If a subclass overwrites ``compute_matrix`` to use a custom signature,
+            this method has to be likewise overwritten.
```

Comment:
Happy for suggestions to make this note clearer

Reply:
Yes they could do that as well. It's a good point, because we could have `matrix` defined, but not `compute_matrix`. 

I think we should say a representation is defined if the `matrix` method is defined, and never rely on `compute_matrix`, which is more a convenience function or helper? This could be fixed by the "registration system"!

Reply:
Yes! I'm in favour if it makes things easier for developers.

Reply:
What might be a reasonable situation where this could happen?

I see this as an avoid unless absolutely necessary situation.  Especially given how much Operator addition and maintenance is done by first-time contributors, we might not want to be suggesting something that isn't recommended except when absolutely necessary,